design-proposals/klip-24-key-column-semantics-in-queries.md

```diff
@@ -0,0 +1,506 @@
+
+### Removal of non-standard Aliasing
+
+The [allow any key name](https://github.com/confluentinc/ksql/issues/3536) feature has introduced
+the ability to provide an alias within a `PARTITION BY`, `GROUP BY` and `JOIN` clause. These are
+non-standard, but required in the current model to allow users to name the key column in the schema.
+For example:
+
+```sql
+-- without alias:
+CREATE STREAM OUTPUT AS SELECT ID FROM INPUT PARTITION BY V0 - V1;
+-- resulting schema: KSQL_COL_0 INT KEY, ID;
+-- note the system generated column name.
+
+-- with alias
+CREATE STREAM OUTPUT AS SELECT ID FROM INPUT PARTITION BY V0 - V1 AS NEW_KEY;
+-- resulting schema: NEW_KEY INT KEY, ID;
+```
+
+However, the same functionality can be achieved using standard sql if the key column is required
+in the projection, for example:
+
+```sql
+CREATE STREAM OUTPUT AS SELECT V0 - V1 AS NEW_KEY, ID FROM INPUT PARTITION BY V0 - V1;
+-- resulting schema: NEW_KEY INT KEY, ID;
+```
+
+### Any key name
+
+Persistent queries failing if the projection contained the key column existed before the work to
+[allow any key name](https://github.com/confluentinc/ksql/issues/3536), though this work also
+exacerbated the situation, as the key column is no longer always called `ROWKEY`. Instead, users can
+pick any name for the key column, and the key column takes on the name of any `PARTITION BY`,
+`GROUP BY` or `JOIN ON` on a single column. This causes seemingly correct and common patterns to
+fail with duplicate column errors:
+
+```sql
+-- Before 'any key name':
+CREATE STREAM INPUT (ROWKEY INT KEY, V0 INT, V1 INT) WITH (...);
+CREATE TABLE OUTPUT AS SELECT V0, COUNT(*) AS COUNT FROM INPUT GROUP BY V0;
+-- resulting schema: ROwKEY INT KEY, V0 INT, COUNT BIGINT
+-- Note: ROWKEY & V0 store the same data, which isn't ideal.
+
+-- vs --
+
+-- After 'any key name':
+CREATE STREAM INPUT (ID INT KEY, V0 INT, V1 INT) WITH (...);
+CREATE TABLE OUTPUT AS SELECT V0, COUNT(*) AS COUNT FROM INPUT GROUP BY V0;
+-- fails with duplicate column error on 'V0'.
+
+-- Note: the equivalent query would have failed before the any-key-name work:
+CREATE TABLE OUTPUT AS SELECT ROWKEY, COUNT(*) AS COUNT FROM INPUT GROUP BY V0;
+-- fails with duplicate column error on 'ROWKEY'.
+```
+
+As you can see from above, the common pattern of selecting the group by key and the aggregate fails.
+To 'fix' the query the user must remove `V0` from the projection - which is counter intuitive, as
+the user wishes this column in the result.
+
+We propose that this query should work, without any modification, and without storing duplicate
+data.
+
+### Requiring the key column in a projection in persistent queries
+
+By flipping the semantics so that the key column is not _implicitly_ copied across in persistent
+queries we place the burden on the user to _explicitly_ specify it in the projection.
+
+Where as it is possible to have a transient query without the key column in the projection, as a
```

Comment:
Personally, I don't buy this argument. The result of a query that returns a _table changelog_ stream (even if it's a stream) should contain the PK IMHO and should be mandatory to have the PK in the `SELECT` clause... (the `EMIT CHANGE` should really just be an "addendum" to the query without altering what is allowed or not allowed). This difference in behavior is rather confusion and also does not allow to use those push queries (ie, that remove the PK) directly as persistent `CT` query, but require a rewrite. IMHO, it would be better to align the semantics.

Reply:
I can see your point of view. A push query of a table without the PK wouldn't be of much use, as the consumer of the data would never know if a specific row was an insert, update or a delete.

The same is not true of a push query of a stream, where, as you know, the key does not infer any insert/update/delete semantics. 

However, push queries are used by humans as much as by machines.  In the CLI a user may just want to see the data they're interested in scrolling across the screen.  They maybe only interested in a single column. Maybe they're looking for a certain value or pattern, or who knows what.  But the point is, I don't see any benefit from forcing a user to select the PK column(s) if they don't want them cluttering their screen.

So I'm afraid I'm against forcing PKs into transient query projections.

cc @derekjn for a product view.

Reply:
> The same is not true of a push query of a stream, where, as you know, the key does not infer any insert/update/delete semantics.

Sure. But as you know, I personally find our current query categorization confusing and not helpful, and applying a "rule" to a query that returns a fact stream to a query that returns a changelog stream, because both a in the same category of "push queries", is not a good argument IMHO.

Reply:
That's not really my argument :p

I'm saying we should let users select whatever they want in the CLI and via other means.  If they don't want the key, that's their business, not ours.

Reply:
I am most worried about _consistency_:

Currently, the following is supported:
```sql
// transient
SELECT count(foo) FROM stream GROUP BY bar EMIT CHANGES;
// persistent
CREATE TABLE table AS SELECT count(foo) FROM GROUP BY bar;
```
Note, that the actual query `SELECT count(foo) FROM stream GROUP BY bar` is exactly the same for both cases.  This allows to deploy a transient query without any change as a persistent query.

With your proposal, the following inconsistency arrises:
```sql
// transient
SELECT count(foo) FROM stream GROUP BY bar EMIT CHANGES;
// fails -> PK missing
CREATE TABLE table AS SELECT count(foo) FROM GROUP BY bar;

// need to rewrite the query and add the PK `bar` to make it work
CREATE TABLE table AS SELECT bar, count(foo) FROM GROUP BY bar;
```

This seems to be issue with regard to UX, because transient queries and persistent queries would not support "different SQL".

Hence, if we require the PK in the transient query, we not only guide users to write "non-sense" queries, but also avoid this inconsistency:
```
// transient -> fails
SELECT count(foo) FROM stream GROUP BY bar EMIT CHANGES;
// persistent -> fails
CREATE TABLE table AS SELECT count(foo) FROM GROUP BY bar;

// work both
// transient
SELECT bar, count(foo) FROM stream GROUP BY bar EMIT CHANGES;
// persistent
CREATE TABLE table AS SELECT bar, count(foo) FROM GROUP BY bar;
```

Reply:
> However, push queries are used by humans as much as by machines.

I think it's important to consider one of the most common contexts within which humans use  transient push queries: iterating on queries before creating them as persistent queries. For this reason, I'm in favor of requiring the PK for push queries. Being able to take a push query (verbatim) and turn it into a persistent query is pretty important from a UX perspective IMO.

I totally understand the reasoning behind not requiring a PK for transient queries, but personally I just don't feel that it would offset the downside of the aforementioned UX inconsistency. I think that predictability and consistency are more important than elegance in this case.

Reply:
IMHO this is total madness. 

Consider yourself working with a traditional db system.  You're selecting some data from a table, which has a primary key.  For whatever reason, you're interested seeing only some of the columns, and that doesn't include the primary key columns.  Does the db stop you doing this? Of course not!

Can you imagine: 

```sql
-- Given table K1, K2, K3 => V0, V1, V2

-- I just want to see column V0:
SELECT V0 FROM S;
-- ERROR: Computer says no; you need to include K1, K2, K3 in your projection, even though you're not interested in them.
```

Madness!

Also, what about a query such as:

```sql
-- how many rows have a non-null V1?
SELECT COUNT(V1) FROM S;
```

Sure, this doesn't work at the moment in ksql, but it's a requested feature.  The result doesn't have a key to include!

My point is there are valid, useful, push queries that users will want to run that don't return data with a primary key.  Obviously, such queries can't be converted into a persistent query that builds a MV/table, as such MVs _currently_ require a primary key.  But doesn't mean we shouldn't allow the user to execute them.  

I strongly disagree that any and every push query must be convertible into a persistent query, (while we require MVs to have a primary key).

Reply:
Another example:

```sql
SELECT COUNT(*) FROM S GROUP BY 1;
```

This is valid right now and will return the number of rows in the table.  Are we saying this should fail, because the key is not in the projection? We'd require them to do what?

```sql
SELECT 1, COUNT(*) FROM S GROUP BY 1;
```

Why????

Reply:
I think the different mind set really comes from "pull" vs "push" queries. For "pull" queries, what you say is more acceptable. However, for "push" queries, as you receive multiple records per row, without the key column you cannot make any sense out of it. What doos
```
SELECT COUNT(*) FROM S GROUP BY a EMIT CHANGE;
```
give you? If you get two updated `3`, `5` you don't know if there are 2 rows with value 3 and 5 respectively, or if there is one row that was updated from 3 to 5.

For a pull query it would of course be different: a table with two rows would return 3 and 5 (what might still be useful even if you don't know the PK of each key) while a table with only one row would either return 3 (if you query before the update) or 5 (if you query after the update).

And I also believe, the the lions shares of push queries is during development to figure out the right persistent query one wants to delpoy.

Reply:
After much discussion, this will be fixed by: https://github.com/confluentinc/ksql/issues/5303

Reply:
In the end it's subjective which one is better/worse. Flexibility vs consistency.

As this KLIP is about _persistent queries_ it seem ok to not change _transient_ queries for now.